spark/src/main/scala/org/apache/spark/sql/delta/commands/alterDeltaTableCommands.scala

```diff
@@ -1005,17 +1005,26 @@ case class AlterTableAddConstraintDeltaCommand(
           (Constraints.checkConstraintPropertyName(name) -> exprText)
       )
 
-      val expr = sparkSession.sessionState.sqlParser.parseExpression(exprText)
-      if (expr.dataType != BooleanType) {
-        throw DeltaErrors.checkConstraintNotBoolean(name, exprText)
+      val df = txn.snapshot.deltaLog.createDataFrame(txn.snapshot, txn.filterFiles())
+      val unresolvedExpr = sparkSession.sessionState.sqlParser.parseExpression(exprText)
+
+      try {
+        df.where(new Column(unresolvedExpr)).queryExecution.analyzed
+      } catch {
+        case a: AnalysisException
+            if a.errorClass.contains("DATATYPE_MISMATCH.FILTER_NOT_BOOLEAN") =>
+          throw DeltaErrors.checkConstraintNotBoolean(name, exprText)
+        case a: AnalysisException =>
+          // Strip out the context of the DataFrame that was used to analyze the expression.
+          throw a.copy(context = Array.empty)
```

Comment:
Thank you for the suggestion. This is indeed a fair bit simpler. Unfortunately this will result in a `_LEGACY_ERROR_TEMP_DELTA_0012` error, instead of the much clearer `UNRESOLVED_COLUMN.WITH_SUGGESTION` exception that we get with the current code.